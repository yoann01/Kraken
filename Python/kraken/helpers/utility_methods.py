from kraken.core.maths.math_object import MathObject
from kraken.core.maths.vec3 import Vec3
from kraken.core.maths.xfo import Xfo
from kraken.core.maths.quat import Quat

from kraken.core.maths import decodeValue


def logHierarchy(kObject):
    """Traverses the given Kraken hierarchy and logs the names of all the objects.

    Args:
        Object: The object to start logging on.

    """

    print kObject.getDecoratedPath()
    for i in xrange(kObject.getNumChildren()):
        child = kObject.getChildByIndex(i)
        logHierarchy(child)


def __convertFromJSON(jsonData):

    if type(jsonData) is list:
        newList = []
        for item in jsonData:
            newList.append(__convertFromJSON(item))
        return newList
    elif type(jsonData) is dict:
        if '__mathObjectClass__' in jsonData.keys():
            return decodeValue(jsonData)
        for key, value in jsonData.iteritems():
            jsonData[key] = __convertFromJSON(value)
    return jsonData


def prepareToLoad(jsonData):
    """Prepares the json data for loading into kraken.

    Args:
        jsonData (dict): The JSON data to be prepared.

    Returns:
        dict: The prepared JSON hierarchy.

    """

    return __convertFromJSON(jsonData)


def __convertToJSON(jsonData):

    if isinstance(jsonData, MathObject):
        return jsonData.jsonEncode()
    elif type(jsonData) is list:
        newList = []
        for item in jsonData:
            newList.append(__convertToJSON(item))
        return newList
    elif type(jsonData) is dict:
        newDict = {}
        for key, value in jsonData.iteritems():
            newDict[key] = __convertToJSON(value)
        return newDict
    return jsonData


def prepareToSave(jsonData):
    """Prepares the json data for serialization.

    Args:
        jsonData (dict): The JSON data to be prepared.

    Returns:
        dict: The prepared JSON hierarchy.

    """

    return __convertToJSON(jsonData)


def __mirrorData(jsonData, plane):

    if isinstance(jsonData, Vec3):
        newVec3 = Vec3(jsonData)
        if plane == 0:
            newVec3.x = -newVec3.x
        elif plane == 1:
            newVec3.y = -newVec3.y
        elif plane == 2:
            newVec3.z = -newVec3.z
        return newVec3

    if isinstance(jsonData, Quat):
        newQuat = Quat(jsonData)
        newQuat.mirror(plane)
        return newQuat

    elif isinstance(jsonData, Xfo):
        newXfo = Xfo(jsonData)
        if plane == 0:
            newXfo.tr.x = -newXfo.tr.x
        elif plane == 1:
            newXfo.tr.y = -newXfo.tr.y
        elif plane == 2:
            newXfo.tr.z = -newXfo.tr.z

        newXfo.ori.mirror(plane)
        return newXfo

    elif type(jsonData) is list:
        newList = []
        for item in jsonData:
            newList.append(__mirrorData(item, plane))
        return newList

    elif type(jsonData) is dict:
        newDict = {}
        for key, value in jsonData.iteritems():
            newDict[key] = __mirrorData(value, plane)
        return newDict

    return jsonData


def mirrorData(jsonData, plane):
    """Prepares the json data for serialization.

    Args:
        jsonData (dict): The JSON data to be prepared.
        plane (int): The plane to mirror across.

    Returns:
        dict: The prepared JSON hierarchy.

    """

    return __mirrorData(jsonData, plane)
