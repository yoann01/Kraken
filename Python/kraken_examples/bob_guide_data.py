"""Example data that can be loaded onto a guide rig."""

from kraken.core.maths import Vec2, Vec3, Euler, Quat, Xfo


bob_guide_data = {
    "components":[
        {
            "class": "kraken_components.biped.spine_component.SpineComponentGuide",
            'cogPosition': Vec3(0.0, 11.1351, -0.1382),
            'spine01Position': Vec3(0.0, 11.1351, -0.1382),
            'spine02Position': Vec3(0.0, 11.8013, -0.1995),
            'spine03Position': Vec3(0.0, 12.4496, -0.3649),
            'spine04Position': Vec3(0.0, 13.1051, -0.4821),
            'numDeformers': 6
        },
        {
            "class": "kraken_components.biped.neck_component.NeckComponentGuide",
            "neckPosition": Vec3(0.0, 16.5572, -0.6915),
            "neckEndPosition": Vec3(0.0, 17.4756, -0.421)
        },
        {
            "class": "kraken_components.biped.head_component.HeadComponentGuide",
            "headPosition": Vec3(0.0, 17.4756, -0.421),
            "headEndPosition": Vec3(0.0, 19.5, -0.421),
            "eyeLeftPosition": Vec3(0.3497, 18.0878, 0.6088),
            "eyeRightPosition": Vec3(-0.3497, 18.0878, 0.6088),
            "jawPosition": Vec3(0.0, 17.613, -0.2731)
        },
        {
            "class": "kraken_components.biped.clavicle_component.ClavicleComponentGuide",
            "name": "Clavicle",
            "location": "L",
            "clavicleXfo": Xfo(Vec3(0.1322, 15.403, -0.5723)),
            "clavicleUpVXfo": Xfo(Vec3(0.0, 1.0, 0.0)),
            "clavicleEndXfo": Xfo(Vec3(2.27, 15.295, -0.753))
        },
        {
            "class": "kraken_components.biped.clavicle_component.ClavicleComponentGuide",
            "name": "Clavicle",
            "location": "R",
            "clavicleXfo": Xfo(Vec3(-0.1322, 15.403, -0.5723)),
            "clavicleUpVXfo": Xfo(Vec3(0.0, 1.0, 0.0)),
            "clavicleEndXfo": Xfo(Vec3(-2.27, 15.295, -0.753))
        },
        {
            "class": "kraken_components.biped.arm_component.ArmComponentGuide",
            "name": "Arm",
            "location": "L",
            "bicepXfo": Xfo(Vec3(2.27, 15.295, -0.753)),
            "forearmXfo": Xfo(Vec3(5.039, 13.56, -0.859)),
            "wristXfo": Xfo(Vec3(7.1886, 12.2819, 0.4906)),
            "handXfo": Xfo(tr=Vec3(7.1886, 12.2819, 0.4906),
                           ori=Quat(Vec3(-0.0865, -0.2301, -0.2623), 0.9331)),
            "bicepFKCtrlSize": 1.75,
            "forearmFKCtrlSize": 1.5
        },
        {
            "class": "kraken_components.biped.arm_component.ArmComponentGuide",
            "name": "Arm",
            "location": "R",
            "bicepXfo": Xfo(Vec3(-2.27, 15.295, -0.753)),
            "forearmXfo": Xfo(Vec3(-5.039, 13.56, -0.859)),
            "wristXfo": Xfo(Vec3(-7.1886, 12.2819, 0.4906)),
            "handXfo": Xfo(tr=Vec3(-7.1886, 12.2819, 0.4906),
                           ori=Quat(Vec3(-0.2301, -0.0865, -0.9331), 0.2623)),
            "bicepFKCtrlSize": 1.75,
            "forearmFKCtrlSize": 1.5
        },
        {
            "class": "kraken_components.biped.leg_component.LegComponentGuide",
            "name": "Leg",
            "location": "L",
            "femurXfo": Xfo(Vec3(0.9811, 9.769, -0.4572)),
            "kneeXfo": Xfo(Vec3(1.4488, 5.4418, -0.5348)),
            "ankleXfo": Xfo(Vec3(1.841, 1.1516, -1.237)),
            "toeXfo": Xfo(Vec3(1.85, 0.4, 0.25)),
            "toeTipXfo": Xfo(Vec3(1.85, 0.4, 1.5))
        },
        {
            "class": "kraken_components.biped.leg_component.LegComponentGuide",
            "name": "Leg",
            "location": "R",
            "femurXfo": Xfo(Vec3(-0.9811, 9.769, -0.4572)),
            "kneeXfo": Xfo(Vec3(-1.4488, 5.4418, -0.5348)),
            "ankleXfo": Xfo(Vec3(-1.841, 1.1516, -1.237)),
            "toeXfo": Xfo(Vec3(-1.85, 0.4, 0.25)),
            "toeTipXfo": Xfo(Vec3(-1.85, 0.4, 1.5))
        }
    ],
    "connections": [
        {
            "_comment": "Neck to Spine",
            "source": "spine:M.spineEnd",
            "target": "neck:M.neckBase"
        },
        {
            "_comment": "Head to Neck",
            "source": "neck:M.neckEnd",
            "target": "head:M.headBase"
        },
        {
            "_comment": "LClavicle to Spine",
            "source": "spine:M.spineEnd",
            "target": "Clavicle:L.spineEnd"
        },
        {
            "_comment": "LArm to LClavicle",
            "source": "Clavicle:L.clavicleEnd",
            "target": "Arm:L.root"
        },
        {
            "_comment": "RArm to RClavicle",
            "source": "Clavicle:R.clavicleEnd",
            "target": "Arm:R.root"
        },
        {
            "_comment": "RClavicle to Spine",
            "source": "spine:M.spineEnd",
            "target": "Clavicle:R.spineEnd"
        },
        {
            "_comment": "LLeg To Pelvis Connections",
            "source": "spine:M.pelvis",
            "target": "Leg:L.pelvisInput"
        },
        {
            "_comment": "RLeg To Pelvis Connections",
            "source": "spine:M.pelvis",
            "target": "Leg:R.pelvisInput"
        }
    ]
}
