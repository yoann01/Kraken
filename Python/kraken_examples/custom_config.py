

from kraken.core.configs.config import Config

class CustomConfig(Config):
    """Custom Config Example.

    This demonstrates to users how they can customize only the name template
    portion of a Config when sub-classed off of the base Config.

    """

    def __init__(self):
        super(CustomConfig, self).__init__()

    # ======================
    # Name Template Methods
    # ======================
    def initNameTemplate(self):
        """Initializes the name template.

        Returns:
            dict: name template.

        """

        nameTemplate = super(CustomConfig, self).initNameTemplate()
        nameTemplate["formats"] =  {
                                   "Container": ["name"],
                                   "Layer": ["container", "sep", "name"],
                                   "ComponentGroup": ["location", "sep", "name", "sep", "type"],
                                   "default": ["location", "sep", "component", "sep", "name", "sep", "type"],
                                  }

        return nameTemplate



from kraken.core.kraken_system import KrakenSystem
ks = KrakenSystem.getInstance()
ks.registerConfig(CustomConfig)
