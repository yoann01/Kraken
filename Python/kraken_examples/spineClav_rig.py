from kraken.core.maths import Vec3, Quat, Xfo

from kraken.core.objects.container import Container

from kraken_components.biped.clavicle_component import ClavicleComponentGuide, ClavicleComponentRig
from kraken_components.biped.spine_component import SpineComponentRig

from kraken.core.profiler import Profiler


class SpineClavRig(Container):
    """Spine Clav Rig

    This example demonstrates how users can create scripted rigs that both load
    data onto a Rig class, and also loads data onto a guide class and pulls the
    data off to the associated Rig class, then builds.

    """

    def __init__(self, name):

        Profiler.getInstance().push("Construct SpineClavRig:" + name)
        super(SpineClavRig, self).__init__(name)

        # Add Components to Layers
        spineComponent = SpineComponentRig("spine", self)
        spineComponent.loadData(data={
            'cogPosition': Vec3(0.0, 11.1351, -0.1382),
            'spine01Position': Vec3(0.0, 11.1351, -0.1382),
            'spine02Position': Vec3(0.0, 11.8013, -0.1995),
            'spine03Position': Vec3(0.0, 12.4496, -0.3649),
            'spine04Position': Vec3(0.0, 13.1051, -0.4821),
            'numDeformers': 4
        })

        clavicleLeftComponentGuide = ClavicleComponentGuide("clavicleGuide")
        clavicleLeftComponentGuide.loadData({
            "location": "L",
            "clavicleXfo": Xfo(Vec3(0.1322, 15.403, -0.5723)),
            "clavicleUpVXfo": Xfo(Vec3(0.0, 1.0, 0.0)),
            "clavicleEndXfo": Xfo(Vec3(2.27, 15.295, -0.753))
        })

        clavicleLeftComponent = ClavicleComponentRig("clavicle", self)
        clavicleLeftComponent.loadData(data=clavicleLeftComponentGuide.getRigBuildData())

        # Clavicle to Spine
        vertebraeOutputs = spineComponent.getOutputByName('spineVertebrae')
        clavicleLeftSpineEndInput = clavicleLeftComponent.getInputByName('spineEnd')
        clavicleLeftSpineEndInput.setConnection(vertebraeOutputs, index = 2)

        Profiler.getInstance().pop()
