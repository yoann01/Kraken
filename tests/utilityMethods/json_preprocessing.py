import json

from kraken_examples.bob_guide_data import bob_guide_data
from kraken.helpers.utility_methods import prepareToSave, prepareToLoad


pureJSON = prepareToSave(bob_guide_data)
str1 = json.dumps(pureJSON, indent=2)
print str1

bob_guide_data = prepareToLoad(pureJSON)

# Check that we can preprocess a data struct 2X
bob_guide_data = prepareToLoad(bob_guide_data)

# Re export to pure json
pureJSON = prepareToSave(bob_guide_data)
str2 = json.dumps(pureJSON, indent=2)
print str1 == str2