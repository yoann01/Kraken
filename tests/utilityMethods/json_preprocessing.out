{
  "connections": [
    {
      "source": "spine:M.spineEnd", 
      "_comment": "Neck to Spine", 
      "target": "neck:M.neckBase"
    }, 
    {
      "source": "neck:M.neckEnd", 
      "_comment": "Head to Neck", 
      "target": "head:M.headBase"
    }, 
    {
      "source": "spine:M.spineEnd", 
      "_comment": "LClavicle to Spine", 
      "target": "Clavicle:L.spineEnd"
    }, 
    {
      "source": "Clavicle:L.clavicleEnd", 
      "_comment": "LArm to LClavicle", 
      "target": "Arm:L.root"
    }, 
    {
      "source": "Clavicle:R.clavicleEnd", 
      "_comment": "RArm to RClavicle", 
      "target": "Arm:R.root"
    }, 
    {
      "source": "spine:M.spineEnd", 
      "_comment": "RClavicle to Spine", 
      "target": "Clavicle:R.spineEnd"
    }, 
    {
      "source": "spine:M.pelvis", 
      "_comment": "LLeg To Pelvis Connections", 
      "target": "Leg:L.pelvisInput"
    }, 
    {
      "source": "spine:M.pelvis", 
      "_comment": "RLeg To Pelvis Connections", 
      "target": "Leg:R.pelvisInput"
    }
  ], 
  "components": [
    {
      "spine03Position": {
        "y": 12.449600219726562, 
        "x": 0.0, 
        "__mathObjectClass__": "Vec3", 
        "z": -0.36489999294281006
      }, 
      "spine01Position": {
        "y": 11.135100364685059, 
        "x": 0.0, 
        "__mathObjectClass__": "Vec3", 
        "z": -0.13819999992847443
      }, 
      "numDeformers": 6, 
      "cogPosition": {
        "y": 11.135100364685059, 
        "x": 0.0, 
        "__mathObjectClass__": "Vec3", 
        "z": -0.13819999992847443
      }, 
      "spine02Position": {
        "y": 11.801300048828125, 
        "x": 0.0, 
        "__mathObjectClass__": "Vec3", 
        "z": -0.19949999451637268
      }, 
      "class": "kraken_components.biped.spine_component.SpineComponentGuide", 
      "spine04Position": {
        "y": 13.10509967803955, 
        "x": 0.0, 
        "__mathObjectClass__": "Vec3", 
        "z": -0.4821000099182129
      }
    }, 
    {
      "neckPosition": {
        "y": 16.557199478149414, 
        "x": 0.0, 
        "__mathObjectClass__": "Vec3", 
        "z": -0.6915000081062317
      }, 
      "class": "kraken_components.biped.neck_component.NeckComponentGuide", 
      "neckEndPosition": {
        "y": 17.47559928894043, 
        "x": 0.0, 
        "__mathObjectClass__": "Vec3", 
        "z": -0.42100000381469727
      }
    }, 
    {
      "jawPosition": {
        "y": 17.613000869750977, 
        "x": 0.0, 
        "__mathObjectClass__": "Vec3", 
        "z": -0.27309998869895935
      }, 
      "eyeLeftPosition": {
        "y": 18.087799072265625, 
        "x": 0.349700003862381, 
        "__mathObjectClass__": "Vec3", 
        "z": 0.6087999939918518
      }, 
      "headPosition": {
        "y": 17.47559928894043, 
        "x": 0.0, 
        "__mathObjectClass__": "Vec3", 
        "z": -0.42100000381469727
      }, 
      "headEndPosition": {
        "y": 19.5, 
        "x": 0.0, 
        "__mathObjectClass__": "Vec3", 
        "z": -0.42100000381469727
      }, 
      "class": "kraken_components.biped.head_component.HeadComponentGuide", 
      "eyeRightPosition": {
        "y": 18.087799072265625, 
        "x": -0.349700003862381, 
        "__mathObjectClass__": "Vec3", 
        "z": 0.6087999939918518
      }
    }, 
    {
      "name": "Clavicle", 
      "clavicleUpVXfo": {
        "sc": {
          "y": 1.0, 
          "x": 1.0, 
          "__mathObjectClass__": "Vec3", 
          "z": 1.0
        }, 
        "__mathObjectClass__": "Xfo", 
        "tr": {
          "y": 1.0, 
          "x": 0.0, 
          "__mathObjectClass__": "Vec3", 
          "z": 0.0
        }, 
        "ori": {
          "__mathObjectClass__": "Quat", 
          "w": 1.0, 
          "v": {
            "y": 0.0, 
            "x": 0.0, 
            "__mathObjectClass__": "Vec3", 
            "z": 0.0
          }
        }
      }, 
      "clavicleEndXfo": {
        "sc": {
          "y": 1.0, 
          "x": 1.0, 
          "__mathObjectClass__": "Vec3", 
          "z": 1.0
        }, 
        "__mathObjectClass__": "Xfo", 
        "tr": {
          "y": 15.295000076293945, 
          "x": 2.2699999809265137, 
          "__mathObjectClass__": "Vec3", 
          "z": -0.753000020980835
        }, 
        "ori": {
          "__mathObjectClass__": "Quat", 
          "w": 1.0, 
          "v": {
            "y": 0.0, 
            "x": 0.0, 
            "__mathObjectClass__": "Vec3", 
            "z": 0.0
          }
        }
      }, 
      "location": "L", 
      "class": "kraken_components.biped.clavicle_component.ClavicleComponentGuide", 
      "clavicleXfo": {
        "sc": {
          "y": 1.0, 
          "x": 1.0, 
          "__mathObjectClass__": "Vec3", 
          "z": 1.0
        }, 
        "__mathObjectClass__": "Xfo", 
        "tr": {
          "y": 15.402999877929688, 
          "x": 0.13220000267028809, 
          "__mathObjectClass__": "Vec3", 
          "z": -0.5723000168800354
        }, 
        "ori": {
          "__mathObjectClass__": "Quat", 
          "w": 1.0, 
          "v": {
            "y": 0.0, 
            "x": 0.0, 
            "__mathObjectClass__": "Vec3", 
            "z": 0.0
          }
        }
      }
    }, 
    {
      "name": "Clavicle", 
      "clavicleUpVXfo": {
        "sc": {
          "y": 1.0, 
          "x": 1.0, 
          "__mathObjectClass__": "Vec3", 
          "z": 1.0
        }, 
        "__mathObjectClass__": "Xfo", 
        "tr": {
          "y": 1.0, 
          "x": 0.0, 
          "__mathObjectClass__": "Vec3", 
          "z": 0.0
        }, 
        "ori": {
          "__mathObjectClass__": "Quat", 
          "w": 1.0, 
          "v": {
            "y": 0.0, 
            "x": 0.0, 
            "__mathObjectClass__": "Vec3", 
            "z": 0.0
          }
        }
      }, 
      "clavicleEndXfo": {
        "sc": {
          "y": 1.0, 
          "x": 1.0, 
          "__mathObjectClass__": "Vec3", 
          "z": 1.0
        }, 
        "__mathObjectClass__": "Xfo", 
        "tr": {
          "y": 15.295000076293945, 
          "x": -2.2699999809265137, 
          "__mathObjectClass__": "Vec3", 
          "z": -0.753000020980835
        }, 
        "ori": {
          "__mathObjectClass__": "Quat", 
          "w": 1.0, 
          "v": {
            "y": 0.0, 
            "x": 0.0, 
            "__mathObjectClass__": "Vec3", 
            "z": 0.0
          }
        }
      }, 
      "location": "R", 
      "class": "kraken_components.biped.clavicle_component.ClavicleComponentGuide", 
      "clavicleXfo": {
        "sc": {
          "y": 1.0, 
          "x": 1.0, 
          "__mathObjectClass__": "Vec3", 
          "z": 1.0
        }, 
        "__mathObjectClass__": "Xfo", 
        "tr": {
          "y": 15.402999877929688, 
          "x": -0.13220000267028809, 
          "__mathObjectClass__": "Vec3", 
          "z": -0.5723000168800354
        }, 
        "ori": {
          "__mathObjectClass__": "Quat", 
          "w": 1.0, 
          "v": {
            "y": 0.0, 
            "x": 0.0, 
            "__mathObjectClass__": "Vec3", 
            "z": 0.0
          }
        }
      }
    }, 
    {
      "forearmFKCtrlSize": 1.5, 
      "name": "Arm", 
      "bicepXfo": {
        "sc": {
          "y": 1.0, 
          "x": 1.0, 
          "__mathObjectClass__": "Vec3", 
          "z": 1.0
        }, 
        "__mathObjectClass__": "Xfo", 
        "tr": {
          "y": 15.295000076293945, 
          "x": 2.2699999809265137, 
          "__mathObjectClass__": "Vec3", 
          "z": -0.753000020980835
        }, 
        "ori": {
          "__mathObjectClass__": "Quat", 
          "w": 1.0, 
          "v": {
            "y": 0.0, 
            "x": 0.0, 
            "__mathObjectClass__": "Vec3", 
            "z": 0.0
          }
        }
      }, 
      "forearmXfo": {
        "sc": {
          "y": 1.0, 
          "x": 1.0, 
          "__mathObjectClass__": "Vec3", 
          "z": 1.0
        }, 
        "__mathObjectClass__": "Xfo", 
        "tr": {
          "y": 13.5600004196167, 
          "x": 5.039000034332275, 
          "__mathObjectClass__": "Vec3", 
          "z": -0.859000027179718
        }, 
        "ori": {
          "__mathObjectClass__": "Quat", 
          "w": 1.0, 
          "v": {
            "y": 0.0, 
            "x": 0.0, 
            "__mathObjectClass__": "Vec3", 
            "z": 0.0
          }
        }
      }, 
      "wristXfo": {
        "sc": {
          "y": 1.0, 
          "x": 1.0, 
          "__mathObjectClass__": "Vec3", 
          "z": 1.0
        }, 
        "__mathObjectClass__": "Xfo", 
        "tr": {
          "y": 12.281900405883789, 
          "x": 7.188600063323975, 
          "__mathObjectClass__": "Vec3", 
          "z": 0.49059998989105225
        }, 
        "ori": {
          "__mathObjectClass__": "Quat", 
          "w": 1.0, 
          "v": {
            "y": 0.0, 
            "x": 0.0, 
            "__mathObjectClass__": "Vec3", 
            "z": 0.0
          }
        }
      }, 
      "location": "L", 
      "handXfo": {
        "sc": {
          "y": 1.0, 
          "x": 1.0, 
          "__mathObjectClass__": "Vec3", 
          "z": 1.0
        }, 
        "__mathObjectClass__": "Xfo", 
        "tr": {
          "y": 12.281900405883789, 
          "x": 7.188600063323975, 
          "__mathObjectClass__": "Vec3", 
          "z": 0.49059998989105225
        }, 
        "ori": {
          "__mathObjectClass__": "Quat", 
          "w": 0.9330999851226807, 
          "v": {
            "y": -0.23010000586509705, 
            "x": -0.08649999648332596, 
            "__mathObjectClass__": "Vec3", 
            "z": -0.2623000144958496
          }
        }
      }, 
      "class": "kraken_components.biped.arm_component.ArmComponentGuide", 
      "bicepFKCtrlSize": 1.75
    }, 
    {
      "forearmFKCtrlSize": 1.5, 
      "name": "Arm", 
      "bicepXfo": {
        "sc": {
          "y": 1.0, 
          "x": 1.0, 
          "__mathObjectClass__": "Vec3", 
          "z": 1.0
        }, 
        "__mathObjectClass__": "Xfo", 
        "tr": {
          "y": 15.295000076293945, 
          "x": -2.2699999809265137, 
          "__mathObjectClass__": "Vec3", 
          "z": -0.753000020980835
        }, 
        "ori": {
          "__mathObjectClass__": "Quat", 
          "w": 1.0, 
          "v": {
            "y": 0.0, 
            "x": 0.0, 
            "__mathObjectClass__": "Vec3", 
            "z": 0.0
          }
        }
      }, 
      "forearmXfo": {
        "sc": {
          "y": 1.0, 
          "x": 1.0, 
          "__mathObjectClass__": "Vec3", 
          "z": 1.0
        }, 
        "__mathObjectClass__": "Xfo", 
        "tr": {
          "y": 13.5600004196167, 
          "x": -5.039000034332275, 
          "__mathObjectClass__": "Vec3", 
          "z": -0.859000027179718
        }, 
        "ori": {
          "__mathObjectClass__": "Quat", 
          "w": 1.0, 
          "v": {
            "y": 0.0, 
            "x": 0.0, 
            "__mathObjectClass__": "Vec3", 
            "z": 0.0
          }
        }
      }, 
      "wristXfo": {
        "sc": {
          "y": 1.0, 
          "x": 1.0, 
          "__mathObjectClass__": "Vec3", 
          "z": 1.0
        }, 
        "__mathObjectClass__": "Xfo", 
        "tr": {
          "y": 12.281900405883789, 
          "x": -7.188600063323975, 
          "__mathObjectClass__": "Vec3", 
          "z": 0.49059998989105225
        }, 
        "ori": {
          "__mathObjectClass__": "Quat", 
          "w": 1.0, 
          "v": {
            "y": 0.0, 
            "x": 0.0, 
            "__mathObjectClass__": "Vec3", 
            "z": 0.0
          }
        }
      }, 
      "location": "R", 
      "handXfo": {
        "sc": {
          "y": 1.0, 
          "x": 1.0, 
          "__mathObjectClass__": "Vec3", 
          "z": 1.0
        }, 
        "__mathObjectClass__": "Xfo", 
        "tr": {
          "y": 12.281900405883789, 
          "x": -7.188600063323975, 
          "__mathObjectClass__": "Vec3", 
          "z": 0.49059998989105225
        }, 
        "ori": {
          "__mathObjectClass__": "Quat", 
          "w": 0.2623000144958496, 
          "v": {
            "y": -0.08649999648332596, 
            "x": -0.23010000586509705, 
            "__mathObjectClass__": "Vec3", 
            "z": -0.9330999851226807
          }
        }
      }, 
      "class": "kraken_components.biped.arm_component.ArmComponentGuide", 
      "bicepFKCtrlSize": 1.75
    }, 
    {
      "toeTipXfo": {
        "sc": {
          "y": 1.0, 
          "x": 1.0, 
          "__mathObjectClass__": "Vec3", 
          "z": 1.0
        }, 
        "__mathObjectClass__": "Xfo", 
        "tr": {
          "y": 0.4000000059604645, 
          "x": 1.850000023841858, 
          "__mathObjectClass__": "Vec3", 
          "z": 1.5
        }, 
        "ori": {
          "__mathObjectClass__": "Quat", 
          "w": 1.0, 
          "v": {
            "y": 0.0, 
            "x": 0.0, 
            "__mathObjectClass__": "Vec3", 
            "z": 0.0
          }
        }
      }, 
      "name": "Leg", 
      "ankleXfo": {
        "sc": {
          "y": 1.0, 
          "x": 1.0, 
          "__mathObjectClass__": "Vec3", 
          "z": 1.0
        }, 
        "__mathObjectClass__": "Xfo", 
        "tr": {
          "y": 1.1516000032424927, 
          "x": 1.840999960899353, 
          "__mathObjectClass__": "Vec3", 
          "z": -1.2369999885559082
        }, 
        "ori": {
          "__mathObjectClass__": "Quat", 
          "w": 1.0, 
          "v": {
            "y": 0.0, 
            "x": 0.0, 
            "__mathObjectClass__": "Vec3", 
            "z": 0.0
          }
        }
      }, 
      "kneeXfo": {
        "sc": {
          "y": 1.0, 
          "x": 1.0, 
          "__mathObjectClass__": "Vec3", 
          "z": 1.0
        }, 
        "__mathObjectClass__": "Xfo", 
        "tr": {
          "y": 5.441800117492676, 
          "x": 1.448799967765808, 
          "__mathObjectClass__": "Vec3", 
          "z": -0.5347999930381775
        }, 
        "ori": {
          "__mathObjectClass__": "Quat", 
          "w": 1.0, 
          "v": {
            "y": 0.0, 
            "x": 0.0, 
            "__mathObjectClass__": "Vec3", 
            "z": 0.0
          }
        }
      }, 
      "toeXfo": {
        "sc": {
          "y": 1.0, 
          "x": 1.0, 
          "__mathObjectClass__": "Vec3", 
          "z": 1.0
        }, 
        "__mathObjectClass__": "Xfo", 
        "tr": {
          "y": 0.4000000059604645, 
          "x": 1.850000023841858, 
          "__mathObjectClass__": "Vec3", 
          "z": 0.25
        }, 
        "ori": {
          "__mathObjectClass__": "Quat", 
          "w": 1.0, 
          "v": {
            "y": 0.0, 
            "x": 0.0, 
            "__mathObjectClass__": "Vec3", 
            "z": 0.0
          }
        }
      }, 
      "location": "L", 
      "class": "kraken_components.biped.leg_component.LegComponentGuide", 
      "femurXfo": {
        "sc": {
          "y": 1.0, 
          "x": 1.0, 
          "__mathObjectClass__": "Vec3", 
          "z": 1.0
        }, 
        "__mathObjectClass__": "Xfo", 
        "tr": {
          "y": 9.769000053405762, 
          "x": 0.9811000227928162, 
          "__mathObjectClass__": "Vec3", 
          "z": -0.45719999074935913
        }, 
        "ori": {
          "__mathObjectClass__": "Quat", 
          "w": 1.0, 
          "v": {
            "y": 0.0, 
            "x": 0.0, 
            "__mathObjectClass__": "Vec3", 
            "z": 0.0
          }
        }
      }
    }, 
    {
      "toeTipXfo": {
        "sc": {
          "y": 1.0, 
          "x": 1.0, 
          "__mathObjectClass__": "Vec3", 
          "z": 1.0
        }, 
        "__mathObjectClass__": "Xfo", 
        "tr": {
          "y": 0.4000000059604645, 
          "x": -1.850000023841858, 
          "__mathObjectClass__": "Vec3", 
          "z": 1.5
        }, 
        "ori": {
          "__mathObjectClass__": "Quat", 
          "w": 1.0, 
          "v": {
            "y": 0.0, 
            "x": 0.0, 
            "__mathObjectClass__": "Vec3", 
            "z": 0.0
          }
        }
      }, 
      "name": "Leg", 
      "ankleXfo": {
        "sc": {
          "y": 1.0, 
          "x": 1.0, 
          "__mathObjectClass__": "Vec3", 
          "z": 1.0
        }, 
        "__mathObjectClass__": "Xfo", 
        "tr": {
          "y": 1.1516000032424927, 
          "x": -1.840999960899353, 
          "__mathObjectClass__": "Vec3", 
          "z": -1.2369999885559082
        }, 
        "ori": {
          "__mathObjectClass__": "Quat", 
          "w": 1.0, 
          "v": {
            "y": 0.0, 
            "x": 0.0, 
            "__mathObjectClass__": "Vec3", 
            "z": 0.0
          }
        }
      }, 
      "kneeXfo": {
        "sc": {
          "y": 1.0, 
          "x": 1.0, 
          "__mathObjectClass__": "Vec3", 
          "z": 1.0
        }, 
        "__mathObjectClass__": "Xfo", 
        "tr": {
          "y": 5.441800117492676, 
          "x": -1.448799967765808, 
          "__mathObjectClass__": "Vec3", 
          "z": -0.5347999930381775
        }, 
        "ori": {
          "__mathObjectClass__": "Quat", 
          "w": 1.0, 
          "v": {
            "y": 0.0, 
            "x": 0.0, 
            "__mathObjectClass__": "Vec3", 
            "z": 0.0
          }
        }
      }, 
      "toeXfo": {
        "sc": {
          "y": 1.0, 
          "x": 1.0, 
          "__mathObjectClass__": "Vec3", 
          "z": 1.0
        }, 
        "__mathObjectClass__": "Xfo", 
        "tr": {
          "y": 0.4000000059604645, 
          "x": -1.850000023841858, 
          "__mathObjectClass__": "Vec3", 
          "z": 0.25
        }, 
        "ori": {
          "__mathObjectClass__": "Quat", 
          "w": 1.0, 
          "v": {
            "y": 0.0, 
            "x": 0.0, 
            "__mathObjectClass__": "Vec3", 
            "z": 0.0
          }
        }
      }, 
      "location": "R", 
      "class": "kraken_components.biped.leg_component.LegComponentGuide", 
      "femurXfo": {
        "sc": {
          "y": 1.0, 
          "x": 1.0, 
          "__mathObjectClass__": "Vec3", 
          "z": 1.0
        }, 
        "__mathObjectClass__": "Xfo", 
        "tr": {
          "y": 9.769000053405762, 
          "x": -0.9811000227928162, 
          "__mathObjectClass__": "Vec3", 
          "z": -0.45719999074935913
        }, 
        "ori": {
          "__mathObjectClass__": "Quat", 
          "w": 1.0, 
          "v": {
            "y": 0.0, 
            "x": 0.0, 
            "__mathObjectClass__": "Vec3", 
            "z": 0.0
          }
        }
      }
    }
  ]
}
True
